src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs

```diff
@@ -151,6 +151,12 @@ private static Metadata GetMetadataNoThrow(PortableExecutableReference reference
         {
             // We can reuse the index for any given reference as long as it hasn't changed.
             // So our checksum is just the checksum for the PEReference itself.
+            // First see if the value is already in the cache, to avoid an allocation if possible.
+            if (ChecksumCache.TryGetValue(reference, out var cached))
```

Comment:
Due to roslyn code-gen, the closure will be created regardless here, since it's always created at the start of the method.

```
 .method public hidebysig static class Microsoft.CodeAnalysis.Checksum 
          GetMetadataChecksum(class Microsoft.CodeAnalysis.Solution solution,
                              [Microsoft.CodeAnalysis]Microsoft.CodeAnalysis.PortableExecutableReference reference,
                              [netstandard]System.Threading.CancellationToken cancellationToken) cil managed
  {
    // Code size       60 (0x3c)
    .maxstack  3
    .locals init (class Microsoft.CodeAnalysis.FindSymbols.SymbolTreeInfo/'<>c__DisplayClass48_0' V_0,
             class Microsoft.CodeAnalysis.Checksum V_1,
             bool V_2,
             class Microsoft.CodeAnalysis.Checksum V_3)
    IL_0000:  newobj     instance void Microsoft.CodeAnalysis.FindSymbols.SymbolTreeInfo/'<>c__DisplayClass48_0'::.ctor()
    IL_0005:  stloc.0
    IL_0006:  ldloc.0
    IL_0007:  ldarg.0
    IL_0008:  stfld      class Microsoft.CodeAnalysis.Solution Microsoft.CodeAnalysis.FindSymbols.SymbolTreeInfo/'<>c__DisplayClass48_0'::solution
    IL_000d:  ldloc.0
    IL_000e:  ldarg.2
    IL_000f:  stfld      [netstandard]System.Threading.CancellationToken Microsoft.CodeAnalysis.FindSymbols.SymbolTreeInfo/'<>c__DisplayClass48_0'::cancellationToken
    IL_0014:  nop
    IL_0015:  ldarg.1
    IL_0016:  ldloca.s   V_1
    IL_0018:  call       bool Microsoft.CodeAnalysis.Serialization.ChecksumCache::TryGetValue(object,
                                                                                              class Microsoft.CodeAnalysis.Checksum&)
    IL_001d:  stloc.2
    IL_001e:  ldloc.2
    IL_001f:  brfalse.s  IL_0025

    IL_0021:  ldloc.1
    IL_0022:  stloc.3
    IL_0023:  br.s       IL_003a

    IL_0025:  ldarg.1
    IL_0026:  ldloc.0
    IL_0027:  ldftn      instance object Microsoft.CodeAnalysis.FindSymbols.SymbolTreeInfo/'<>c__DisplayClass48_0'::'<GetMetadataChecksum>b__0'(object)
    IL_002d:  newobj     instance void [netstandard]System.Runtime.CompilerServices.ConditionalWeakTable`2/CreateValueCallback<object,object>::.ctor(object,
                                                                                                                                                     native int)
    IL_0032:  call       class Microsoft.CodeAnalysis.Checksum Microsoft.CodeAnalysis.Serialization.ChecksumCache::GetOrCreate(object,
                                                                                                                               [netstandard]System.Runtime.CompilerServices.ConditionalWeakTable`2/CreateValueCallback<object,object>)
    IL_0037:  stloc.3
    IL_0038:  br.s       IL_003a

    IL_003a:  ldloc.3
    IL_003b:  ret
  } // end of method SymbolTreeInfo::GetMetadataChecksum
```

Reply:
So, it may need splitting the `GetOrCreate` call into its own method.

Reply:
Take 2?

Reply:
There's only take 1 here.

Reply:
Can be marked as resolved.